### PR TITLE
Fix some potential infinity errors in Naive Bayes Classifier.

### PR DESCRIPTION
**Pull request by KimSangYeon-DGU**
This is a fix for #2017. I found out `NaiveBayesClassifier` works correctly for a dataset with low dimensions, like `iris_full.dat`, but for the datasets with high dimensions, it didn't. From tracing the source of the problem, I figured out there were three reasons to cause undesirable results. First, in `LogLikelihood()` method, we replaced zero of the `variance` with 1e-50 to make it invertible, but I think we should update all the values. Therefore, I added a parameter, `epsilon`. Second, in `LogLikelihood()`, `std::log(arma::det(arma::diagmat(variances.col(i))))` caused the infinity of log likelihood for high dimensions, so I changed the equation to more computationally tractable one. Lastly, in `Classify` methods with the parameter, `probabilities`, the Log-Sum-Exp operation sometimes became negative infinity when `variance` has small negative values. Thus, I fixed the code to prevent the potential errors.

Below are the results using test data that were 20% from the total, and I used the code and dataset provided by @marcovirgolin 
**[breastwc_full]**
```
mlpack - mean: 0.93, std: 0.02
sklearn - mean: 0.93, std: 0.02
```

**[cylinderbands_full]**
```
mlpack - mean: 0.31, std: 0.05
sklearn - mean: 0.34, std: 0.05
```

**[ecoli_full]**
```
mlpack - mean: 0.49, std: 0.1
sklearn - mean: 0.5, std: 0.1
```

**[imagesegmentation_full]**
```
mlpack - mean: 0.78, std: 0.01
sklearn - mean: 0.78, std: 0.01
```

**[ionosphere_full]**
```
mlpack - mean: 0.87, std: 0.04
sklearn - mean: 0.88, std: 0.04
```

**[iris_full]**
```
mlpack - mean: 0.95, std: 0.04
sklearn - mean: 0.95, std: 0.04
```

**[madelon_full]**
```
mlpack - mean: 0.6, std: 0.02
sklearn - mean: 0.6, std: 0.02
```

**[sonar_full]**
```
mlpack - mean: 0.65, std: 0.06
sklearn - mean: 0.65, std: 0.06
```

**[vowel_full]**
```
mlpack - mean: 0.66, std: 0.03
sklearn - mean: 0.66, std: 0.03
```

**[yeast_full]**
```
mlpack - mean: 0.25, std: 0.05
sklearn - mean: 0.25, std: 0.04
```

### CODE REVIEW

File: src/mlpack/methods/naive_bayes/naive_bayes_classifier_impl.hpp [hunk 1]
```diff
@@ -315,18 +321,24 @@ void NaiveBayesClassifier<ModelMatType>::Classify(
...
   for (size_t j = 0; j < data.n_cols; ++j)
   {
-    logProbX(j) = log(arma::accu(exp(logLikelihoods.col(j))));
-    logLikelihoods.col(j) -= logProbX(j);
+    // To prevent underflow in log of sum of exp of x operation (where x is a
+    // small negative value), we use logsumexp(x - max(x)) + max(x).
+    maxValue = arma::max(logLikelihoods.col(j));
+    logProbX(j) = log(arma::accu(exp(logLikelihoods.col(j) - maxValue))) +
+        maxValue;
+    predictionLogProb.col(j) = logLikelihoods.col(j) - logProbX(j);
```
Comment (zoq) [inline src/mlpack/methods/naive_bayes/naive_bayes_classifier_impl.hpp]: Looks like if we store the result into ``predictionProbs`` right away we don't need ``predictionLogProb``.

Comment (rcurtin) [reply src/mlpack/methods/naive_bayes/naive_bayes_classifier_impl.hpp]: Right, we could just do `predictionProbs.col(j) = arma::exp(logLikelihoods.col(j) - logProbX(j))`.  Also, there's no need to store `logProbX`, I think---we can just use a single value, or even inline the current computation of `logProbX(j)`.

**Comment by zoq** [comment]
Wait one last comment, can you write a test case that uses the dataset from above or another one that shows the same issue, could be helpful in the future to avoid such issues.

**Comment by rcurtin** [comment]
> Wait one last comment, can you write a test case that uses the dataset from above or another one that shows the same issue, could be helpful in the future to avoid such issues.

Oh, good point.  Should be easy, just be possible just to create an 1000-dimensional linearly separable dataset and then expect that NBC is able to properly classify it.

**Comment by KimSangYeon-DGU** [comment]
> Wait one last comment, can you write a test case that uses the dataset from above or another one that shows the same issue, could be helpful in the future to avoid such issues.

Great. Also, I found the last update failed to pass NBCTest, I'll look into it.

**Comment by rcurtin** [comment]
Ah, oops, I approved too early :-D

**Comment by zoq** [comment]
Same, haven't checked the travis result.

**Comment by KimSangYeon-DGU** [comment]
Ahh, sorry. It's my fault. When I'm done, I'll let you know, please wait to merge this in until then.

**Comment by rcurtin** [comment]
@KimSangYeon-DGU have you had a chance to look at this?  We can make it part of the mlpack-3.2.1 patch release in a couple days if we can resolve the issue by then. :+1:

**Comment by KimSangYeon-DGU** [comment]
@rcurtin Sorry for the slow progress. Yeah, I want this work to be included the mlpack-3.2.1 patch release. :)

**Comment by KimSangYeon-DGU** [comment]
Hmm, `GmmTrainDiffKmeansMaxIterationsTest` failed. However, it is not related to this PR.

**Comment by KimSangYeon-DGU** [comment]
@mlpack-jenkins test this please

**Comment by ShikharJ** [comment]
@KimSangYeon-DGU Please feel free to merge this in.

**Comment by KimSangYeon-DGU** [comment]
@ShikharJ Yeah :)

Thanks, everyone :)

**Comment by rcurtin** [comment]
:+1: I added a note to `HISTORY.md` in 89db422a.

**Pull request merged**